src/init.cpp

```diff
@@ -1792,7 +1792,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     });
 
     // Wait for genesis block to be processed
```

Comment:
I drafted a comment, but I think it's wrong so I haven't pushed it yet:

```cpp
    /* Wait for genesis block to be processed. Typically kernel_notifications.m_tip_block
     * has already been set, except in three cases where the genesis block is
     * being connected by the initload thread:
     *
     * 1. first startup with an empty datadir
     * 2. reindex
     * 3. reindex-chainstate
     */
```

I don't think the `initload` thread ever sets `kernel_notifications.m_tip_block`. The thread only deals with indexes derived from `BaseIndex`, which doesn't include `CBlockIndex`.

I'm still confused under what circumstance it would actually be unset.

Reply:
It's clearly doing something though, because the following breaks several functional tests:

```patch
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1792,13 +1792,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     });
 
     // Wait for genesis block to be processed
     {
         WAIT_LOCK(kernel_notifications.m_tip_block_mutex, lock);
-        kernel_notifications.m_tip_block_cv.wait(lock, [&]() EXCLUSIVE_LOCKS_REQUIRED(kernel_notifications.m_tip_block_mutex) {
-            return !kernel_notifications.m_tip_block.IsNull() || ShutdownRequested(node);
-        });
+        Assert(!kernel_notifications.m_tip_block.IsNull());
+        // kernel_notifications.m_tip_block_cv.wait(lock, [&]() EXCLUSIVE_LOCKS_REQUIRED(kernel_notifications.m_tip_block_mutex) {
+        //     return !kernel_notifications.m_tip_block.IsNull() || ShutdownRequested(node);
+        // });
     }
```

Reply:
With some manual testing I indeed get all three conditions from your comment to hit that assert. I'm just confused why.

Reply:
Outside the context of a snapshot, we only call `LoadChainTip()` if `if (!is_coinsview_empty(chainstate)) `:

https://github.com/bitcoin/bitcoin/blob/2638fdb4f934be96b7c798dbac38ea5ab8a6374a/src/node/chainstate.cpp#L147-L150

That explains why these three scenarios are different.

Still looking into where it _is_ set.

Reply:
In the case of fresh data dir or reindex, it's `ActivateBestChain()` that calls `blockTip()` once it sets the genesis block to the tip.

This indeed happens in the `initload` threads, which seems to keep doing this throughout the reindex process. Once all previously stored blocks are done, it exits and the `msghand` thread takes over.

Reply:
Specifically it's `ImportBlocks()` that does this, even if no blocks are passed to it. 

> which seems to keep doing this throughout the reindex process.

In fact it does it _after_ reindex is finished. `ActivateBestChain` connects blocks one by one until it has nothing left to do so that makes sense.

Reply:
Took the comment and expanded it a bit.

Reply:
> In fact it does it after reindex is finished. ActivateBestChain connects blocks one by one until it has nothing left to do so that makes sense.

Just as an aside: ABC is also called [during](https://github.com/bitcoin/bitcoin/blob/2638fdb4f934be96b7c798dbac38ea5ab8a6374a/src/validation.cpp#L5149)  reindexing when encountering the genesis block. That's why we have this slightly weird behavior with a full `-reindex` that normally first reindexes all blocks and then builds the chainstate, but if the user interrupts it during reindexing and  restarts the node without `-reindex`, it will pick up by first connecting all blocks of the previously indexed block files, then reindex the remaining block files and finally connect the blocks from the second batch.

But that's not really relevant to this PR, the comment seems fine now!

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31346#discussion_r1856824817

> it will pick up by first connecting all blocks of the previously indexed block files

I don't think I understand what the slightly weird behavior is, or maybe what is meant by connecting all blocks. I'd expect if node is started without -reindex, `LoadChainTip` should connect all the blocks that were previously connected, and this would just avoid repeating work and not be weird.

Reply:
The "natural" IBD order is to first download all headers, add them to the block tree database, and then download all blocks / connect them when possible (with "connect" I mean `ActivateBestChain()` / `ConnectBlock()`).
During `reindex` we normally do it similarly: Go through all blockfiles, accept block headers to the db, don't connect any block except Genesis, and only in the end call `ActivateBestChain` to connect blocks / rebuild the chainstate (the reindex-chainstate part).

In the case of interrupt / restart, this order is mixed up. Naively I would expect to load the previously saved state (with only Genesis connected), continue to walk through the block files until all headers are loaded, and finally call `ActivateBestChain` to connect all blocks.

But in reality, we will call already `ActivateBestChain` early, when the best header is somewhere mid-chain, connect a first batch of blocks to the chainstate, then reindex the remaining block files, and then connect a second batch.

When I called this "weird" I did so because it mixes up the normal order, but I didn't know of a downside. However, it has a negative impact that @l0rinc pointed out to me (and I believe is working to fix): When we start connecting blocks without having all headers in our block tree db, we don't apply `assumevalid` (because our best header is not above the minchainwork threshold), and are therefore connecting blocks much slower while reindexing the chainstate.

Reply:
Thanks, so is the scenario specifically the case when a node started with `-reindex` is stopped *before* the final ActivateBestChain call? If it is stopped *during* the final ActivateBestChain call it seems like it would be picking up work where it left off when ActivateBestChain was called again, due to `LoadChainTip` setting the previous tip.

> we will call already `ActivateBestChain` early, when the best header is somewhere mid-chain, connect a first batch of blocks to the chainstate, then reindex the remaining block files, and then connect a second batch.

This is referring to the [ActivateBestChain](https://github.com/bitcoin/bitcoin/blob/2638fdb4f934be96b7c798dbac38ea5ab8a6374a/src/validation.cpp#L5149) call in LoadExternalBlockFile call you linked to earlier? And removing this call would fix the weirdness and assumevalid issue? It does seem like it would be nice to remove this call. I'm not sure what the point of connecting the genesis block early but not connecting any other blocks is.

Reply:
> Thanks, so is the scenario specifically the case when a node started with `-reindex` is stopped _before_ the final ActivateBestChain call?

Yes.

>This is referring to the [ActivateBestChain](https://github.com/bitcoin/bitcoin/blob/2638fdb4f934be96b7c798dbac38ea5ab8a6374a/src/validation.cpp#L5149) call in LoadExternalBlockFile call you linked to earlier?

Yes!

>And removing this call would fix the weirdness and assumevalid issue? It does seem like it would be nice to remove this call. I'm not sure what the point of connecting the genesis block early but not connecting any other blocks is.

I don't think we'd want to remove it completely, because in case of a `-reindex` it allows `init` to complete normal startup without having to wait for the import thread to finish reindexing block files (which can take a long time for mainnet). I think that @l0rinc is looking into possible fixes, but without having tested anything my first idea was to only call `ActivateBestChain` there it if the chain is empty, so it only doesn't get called in the restart scenario where we already loaded Genesis during `LoadChainTip`.

Reply:
This is in my backlog, but I would be very happy if someone who already understands this fixes it instead :)
I just noticed it during Core Dev that my perf logs are full of script validations when trying out different scenarios and cancelling a reindex.

Reply:
> This is in my backlog, but I would be very happy if someone who already understands this fixes it instead :)

Ok thanks, I'll try out my suggestion from above and will open a PR if it's as straightforward as I think.

Reply:
#31439